lib/Transforms/IPO/StructFieldAccessInfo.cpp

```diff
@@ -0,0 +1,343 @@
+          DEBUG_WITH_TYPE(DEBUG_TYPE_IR,
+                          dbgs() << "User is an invoke instruction\n");
+          assert(Inst && isa<InvokeInst>(Inst));
+          auto *F = dyn_cast<InvokeInst>(Inst)->getCalledFunction();
+          if (!F || F->isDeclaration()) {
+            // If a struct is passed to an indirect or a function not declared
+            // in the program, we can't analyze it...
+            Eligiblity = false;
+          }
+        }
+        continue;
+      }
+      addFieldAccessFromGEP(Inst);
+    } else if (isa<Operator>(U)) {
+      auto *Inst = dyn_cast<Operator>(U);
+      if (!isa<GEPOperator>(Inst)) {
+        // Only support access struct through GEP for now
+        continue;
+      }
+      addFieldAccessFromGEP(Inst);
+    } else {
+      addStats(StructFieldAccessManager::DebugStats::
+                   DS_UserNotInstructionNorOperator);
+    }
+  }
+}
+
+void StructFieldAccessInfo::analyzeUsersOfStructPointerValue(const Value *V) {
+  // Analyze users of value defined as struct*
+  assert(V->getType()->isPointerTy() &&
+         V->getType()->getPointerElementType()->isPointerTy() &&
+         V->getType()
+             ->getPointerElementType()
+             ->getPointerElementType()
+             ->isStructTy());
+  for (auto *U : V->users()) {
+    DEBUG_WITH_TYPE(DEBUG_TYPE_IR,
+                    dbgs() << "Analyzing user of " << *V << ": " << *U << "\n");
+    if (isa<Instruction>(U)) {
+      auto *Inst = dyn_cast<Instruction>(U);
+      if (Inst->getOpcode() == Instruction::Load) {
+        analyzeUsersOfStructValue(Inst);
+      } else if (Inst->getOpcode() == Instruction::GetElementPtr) {
+        addStats(StructFieldAccessManager::DebugStats::DS_GepUsedOnStructPtr);
+      } else {
+        addStats(
+            StructFieldAccessManager::DebugStats::DS_UnknownUsesOnStructPtr);
+      }
+    } else if (isa<Operator>(U)) {
+      auto *Inst = dyn_cast<Operator>(U);
+      if (Inst->getOpcode() == Instruction::Load) {
+        analyzeUsersOfStructValue(Inst);
+      } else if (Inst->getOpcode() == Instruction::GetElementPtr) {
+        addStats(StructFieldAccessManager::DebugStats::DS_GepUsedOnStructPtr);
+      } else {
+        addStats(
+            StructFieldAccessManager::DebugStats::DS_UnknownUsesOnStructPtr);
+      }
+    } else {
+      addStats(StructFieldAccessManager::DebugStats::
+                   DS_UserNotInstructionNorOperator);
+    }
+  }
+}
+
+void StructFieldAccessInfo::summarizeFunctionCalls() {
+  DEBUG_WITH_TYPE(DEBUG_TYPE_IR, dbgs() << "Summarizes call/invokes "
+                                           "instructions into function "
+                                           "declarations\n");
+  for (auto &it : CallInstFieldAccessMap) {
+    auto *CallSiteInfo = it.second;
+    assert(CallSiteInfo);
+    auto *F = CallSiteInfo->FunctionDeclaration;
+    assert(F);
+    if (FunctionAccessMap.find(F) == FunctionAccessMap.end()) {
```

Comment:
I don't understand what you mean here. The loop is in the order of each CallInst/InvokeInst found in the IR, their function definitions might not be in order. Here I'd like to search and see if the function declaration is already in the map.